[Hub Generated] Review request for Microsoft.Security to add version preview/2023-12-01-preview

Nessya123 opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/Nessya123-security-Microsoft.Security-2023-12-01-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/029a11fd-eaf9-4865-9d77-a16b5ea5eb26)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [ ] New API version. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

ahnj12 committed:
Update readme.md

Reverting global tag to stable GA version for SDK release

Nessya123 committed:
Adds base for updating Microsoft.Security from version preview/2020-01-01-preview to version 2023-12-01-preview

Nessya123 committed:
Updates readme

Nessya123 committed:
Updates API version in new specs and examples

Commit:
Add 2023-12-01 version

Commit:
Remove files

Commit:
change in readme file

Commit:
fix version

Commit:
Check error

Commit:
back to boolean

Commit:
Fix examples

Commit:
Merge branch 'master' into Nessya123-security-Microsoft.Security-2023-12-01-preview

Commit:
some more fixes

Commit:
Fix version in automation examples

Commit:
.

Commit:
fix

Commit:
fix

Commit:
fixes

Commit:
.

Commit:
fixes

Commit:
more fixes

Nessya123 commented:
/azp run unifiedPipeline

Commit:
more fixes

Commit:
.

Commit:
.

Commit:
.

raosuhas commented:
Can you please add the base version in  the first commit so that we can do a diff to look at the changes across versions.

Also I notice that your team only releases one or 2 files in a new api version. This is not recommended practice since SDKs generated from these will be incomplete which makes it hard for clients to manage the versions. Do you have a plan in place to release all files together ?

Nessya123 commented on specification/security/resource-manager/Microsoft.Security/preview/2023-12-01-preview/automations.json:
```diff
@@ -0,0 +1,716 @@
+{
+  "swagger": "2.0",
```
Automation resource previous version is 2019-01-01 preview

Nessya123 commented on specification/security/resource-manager/Microsoft.Security/preview/2023-12-01-preview/securityContacts.json:
```diff
@@ -0,0 +1,501 @@
+{
+  "swagger": "2.0",
```
security contact resource previous version is 2020-01-01 preview

Nessya123 commented:
> Can you please add the base version in the first commit so that we can do a diff to look at the changes across versions.
> 
> Also I notice that your team only releases one or 2 files in a new api version. This is not recommended practice since SDKs generated from these will be incomplete which makes it hard for clients to manage the versions. Do you have a plan in place to release all files together ?



> Also I notice that your team only releases one or 2 files in a new api version. This is not recommended practice since SDKs generated from these will be incomplete which makes it hard for clients to manage the versions. Do you have a plan in place to release all files together ?

Added comments of the version of each resource I added in this release.
Regarding your second comment- I'm not sure I understand your ask....

sjanamma commented on specification/security/resource-manager/Microsoft.Security/preview/2023-12-01-preview/securityContacts.json:
```diff
@@ -0,0 +1,501 @@
+          "description": "The security contact's phone number"
+        },
+        "isEnabled": {
+          "type": "boolean",
```
ARM recommends enums over booleans for future proof APIs.

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].

Nessya123 replied:
This is a newly exposed security contact property, but it is already present in our backend and databases, so changing it would be difficult. However, in this particular case, I do not see any additional values that we may need. As I mentioned, our backend uses this property for all old versions, but it is not exposed.

sjanamma commented on specification/security/resource-manager/Microsoft.Security/preview/2023-12-01-preview/automations.json:
```diff
@@ -0,0 +1,716 @@
+      "type": "object",
+      "description": "The security automation model state property bag.",
+      "properties": {
+        "isValid": {
```
ARM recommends enums over booleans for future proof APIs.

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].

Nessya123 replied:
This property already exists. Changing it now will cause a breaking change, which I would like to avoid. Additionally, as mentioned in the previous comment, I cannot see any additional values that we would want to add. this simply indicates whether it is enabled or not, no other states are applicable here

Commit:
Resolve comments

Nessya123 commented on specification/quota/resource-manager/readme.md:
````diff
@@ -26,7 +26,7 @@ These are the global settings for the Quota API.
 ``` yaml
 openapi-type: arm
-tag: package-2023-06-01-preview
+tag: package-2023-02-01
````
No, will revert it.

Commit:
Add automation patch request + add empty patren name

Commit:
some fixes

Commit:
fix path example

Commit:
roll back tags changes

Nessya123 commented:
> Please fix the Swagger linter errors https://github.com/Azure/azure-rest-api-specs/pull/26773/checks?check_run_id=18880310075

Regarding two type of checks that are still remaining:
1. ResourceNameRestriction- on the Automation resource there is no restrictions at all and can be any name the customer wants. (as for the security contact resource, I added the relevant restrictions).
2. OperationsAPIImplementation on the security contact resource- we are currently un support resource group scope of the URL. so, I implemented only subscriptions paths.

@sjanamma I would appreciate your help with approve these checks!

Commit:
try

Commit:
fix readme

visingla-ms commented:
> > Please fix the Swagger linter errors https://github.com/Azure/azure-rest-api-specs/pull/26773/checks?check_run_id=18880310075
> 
> Regarding one check that is still remaining:
> 
> 1. ResourceNameRestriction- on the Automation resource there is no restrictions at all and can be any name the customer wants. (as for the security contact resource, I added the relevant restrictions).
> 
> Regarding this check: UnSupportedPatchProperties - I will appreciate it if you could help me understand how can I avoid this as I'm not sure I understand how can I mark the tracked resource properties like the location as read-only
> 
> @sjanamma I would appreciate your help!

`Regarding this check: UnSupportedPatchProperties - I will appreciate it if you could help me understand how can I avoid this as I'm not sure I understand how can I mark the tracked resource properties like the location as read-only`

You need to define another definition separately for Patch (like AutomationUpdateModel), which doesn't have location property. In fact, just add "tags" property in that definition and the others which can actually be patched by your service.

Commit:
Add AutomationUpdateModel

Commit:
missing }

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Nessya123 commented:
> Plese fix Model Validation errors: https://github.com/Azure/azure-rest-api-specs/pull/26773/checks?check_run_id=19324973436, it seems, you have pattern as "Default" with capital cased first letter, whereas in your examples it is with lower cased first letter.

Thank you @visingla-ms   ! the model validation is completed successfully.
Regrading the LintDiff check there is one check called [RequestBodyMustExistForPutPatch] which I don't understand where I am missing the issue, can u please help me understand it?

Commit:
indention fix

ms-zhenhua commented:
Hey @Nessya123, could you please fix the branch conflicts?

<img width="644" alt="image" src="https://github.com/Azure/azure-rest-api-specs/assets/76987228/bd2e92dc-d743-4370-8a8d-62cd458acc8b">

Commit:
Merge branch 'main' into Nessya123-security-Microsoft.Security-2023-12-01-preview

Commit:
Update readme after merge

Nessya123 commented:
> Hey @Nessya123, could you please fix the branch conflicts?
> 
> <img alt="image" width="644" src="https://private-user-images.githubusercontent.com/76987228/289457296-bd2e92dc-d743-4370-8a8d-62cd458acc8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIyODIyMjksIm5iZiI6MTcwMjI4MTkyOSwicGF0aCI6Ii83Njk4NzIyOC8yODk0NTcyOTYtYmQyZTkyZGMtZDc0My00MzcwLThhOGQtNjJjZDQ1OGFjYzhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjExVDA4MDUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhMTRjNDkwNGM2MjhlODVlZTBjZjM4MjE0MTIzYzA5NDE1MTExN2RjNzEzYWVjYTkxOWNkMjgyY2FlYTAxZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jGBjycrCGssm61C1--fcKTBfYu6x7IZDUtfF-2jcrq8">

Done

raych1 commented:
/azp run

Nessya123 commented:
/pr RequestMerge

surashed commented:
[celebrate]     Sulaiman Abu Rashed reacted to your message:
________________________________
From: Nessya123 ***@***.***>
Sent: Tuesday, March 12, 2024 10:57:32 AM
To: Azure/azure-rest-api-specs ***@***.***>
Cc: Sulaiman Abu Rashed ***@***.***>; Comment ***@***.***>
Subject: Re: [Azure/azure-rest-api-specs] [Hub Generated] Review request for Microsoft.Security to add version preview/2023-12-01-preview (PR #26773)


/pr RequestMerge

—
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/26773#issuecomment-1991364720>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANEUVXGJT5EZCZPSWI5JWMTYX3NRZAVCNFSM6AAAAAA7SRKKS2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJRGM3DINZSGA>.
You are receiving this because you commented.Message ID: ***@***.***>
